### PR TITLE
Support more features on BLE related API.

### PR DESCRIPTION
**Pull request by ko1in1u**
1. Support passing `scanResponse` to BLE advertising.
2. Support passing `scanFilters` and `scanSettings` to BLE scanning.
3. Support BluetoothGatt APIs.
4. Support BluetoothGattServer APIs.

- - -
This change is [Reviewable](https://reviewable.io/reviews/google/mobly-bundled-snippets/167)

### CODE REVIEW

File: src/main/java/com/google/android/mobly/snippet/bundled/BluetoothLeAdvertiserSnippet.java [hunk 1]
```diff
@@ -94,13 +96,20 @@ public BluetoothLeAdvertiserSnippet() {
      *          }
      *     </pre>
      *
+     * @param scanResponse A JSONObject representing a {@link AdvertiseData} object is returned when
```
Comment (xpconanfan) [inline src/main/java/com/google/android/mobly/snippet/bundled/BluetoothLeAdvertiserSnippet.java:99]: the method's signature clearly shows return type is void.
why does it say something will be "returned" here?

this is confusing

**Review by xpconanfan — CHANGES_REQUESTED** [review]

File: src/main/java/com/google/android/mobly/snippet/bundled/BluetoothLeAdvertiserSnippet.java (hunk 1 above)
Comment (ko1in1u) [reply src/main/java/com/google/android/mobly/snippet/bundled/BluetoothLeAdvertiserSnippet.java:99]: The `scanResponse` returned in `BluetoothLeAdvertiser#startAdvertising` is a ScanResponse object that contains information about the advertisement that was started. It is saved to this object, not return to the function.

Comment (ko1in1u) [reply src/main/java/com/google/android/mobly/snippet/bundled/BluetoothLeAdvertiserSnippet.java]: Correct the wording.

Comment (ko1in1u) [reply src/main/java/com/google/android/mobly/snippet/bundled/BluetoothLeAdvertiserSnippet.java]: User can pass a dict when calling 

ex:
```
# Advertise data in json format representing scan response for BLE advertise.
SCAN_RESPONSE = {
    'IncludeDeviceName':
        False,
    'ServiceData': [{
        'UUID': SCAN_RESPONSE_UUID,
        'Data': SCAN_RESPONSE_DATA
    }]
}
ad.mbs.bleStartAdvertising(
      ADVERTISE_SETTINGS,
      ADVERTISE_DATA,
      SCAN_RESPONSE,
)
```

Comment (xpconanfan) [reply src/main/java/com/google/android/mobly/snippet/bundled/BluetoothLeAdvertiserSnippet.java]: I mean where would they get the info in `SCAN_RESPONSE`?

Seems like it's returned from a prior call of some API?

Comment (ko1in1u) [reply src/main/java/com/google/android/mobly/snippet/bundled/BluetoothLeAdvertiserSnippet.java]: Yeah, looks like it is used in Java and might not abe able to reused in Python. Not sure if any user needs this so mark it as an optional.

Comment (xpconanfan) [reply src/main/java/com/google/android/mobly/snippet/bundled/BluetoothLeAdvertiserSnippet.java]: Update the docstring to clarify the usage of the optional arg per offline discussion

Comment (ko1in1u) [reply src/main/java/com/google/android/mobly/snippet/bundled/BluetoothLeAdvertiserSnippet.java]: Updated. Thank you.

**Review by xianyuanjia — CHANGES_REQUESTED** [review]
> Reviewed 3 of 10 files at r1, 1 of 1 files at r3.
> *[Reviewable](https://reviewable.io/reviews/google/mobly-bundled-snippets/167)* status: 4 of 10 files reviewed, 3 unresolved discussions (waiting on @ko1in1u and @xpconanfan)

___
*[`src/main/java/com/google/android/mobly/snippet/bundled/bluetooth/BluetoothGattClientSnippet.java` line 2 at r3](https://reviewable.io/reviews/google/mobly-bundled-snippets/167#-NWs453IAQvQrrkr2s0n:-NWs453IAQvQrrkr2s0o:b-jaekzm) ([raw file](https://github.com/google/mobly-bundled-snippets/blob/cc7a7d918f67ad4b2640b04aed84256c5fd316da/src/main/java/com/google/android/mobly/snippet/bundled/bluetooth/BluetoothGattClientSnippet.java#L2)):*
> ```Java
> /*
>  * Copyright (C) 2017 Google Inc.
> ```

Update the year



_Code quote:_
```Java
2017
```
___
*[`src/main/java/com/google/android/mobly/snippet/bundled/utils/DataHolder.java` line 23 at r3](https://reviewable.io/reviews/google/mobly-bundled-snippets/167#-NWs4WnXDsllI4ZX3XCp:-NWs4WnXDsllI4ZX3XCq:b3tj3qc) ([raw file](https://github.com/google/mobly-bundled-snippets/blob/cc7a7d918f67ad4b2640b04aed84256c5fd316da/src/main/java/com/google/android/mobly/snippet/bundled/utils/DataHolder.java#L23)):*
> ```Java
> 
> /** A holder to hold android objects for snippets. */
> public class DataHolder {
> ```

Name should be more specific if this class is meant only for BluetoothGattCharacteristics.

Also, if this is just internally a HashMap, why not use HashMap directly?



_Code quote:_
```Java
DataHolder
```

**Review by xianyuanjia — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/google/mobly-bundled-snippets/167)* status: 4 of 10 files reviewed, 4 unresolved discussions (waiting on @ko1in1u and @xpconanfan)

___
*[`src/main/java/com/google/android/mobly/snippet/bundled/utils/DataHolder.java` line 2 at r3](https://reviewable.io/reviews/google/mobly-bundled-snippets/167#-NWsBw1l9XaMKwT4jSOB:-NWsBw1l9XaMKwT4jSOC:b-4asxxb) ([raw file](https://github.com/google/mobly-bundled-snippets/blob/cc7a7d918f67ad4b2640b04aed84256c5fd316da/src/main/java/com/google/android/mobly/snippet/bundled/utils/DataHolder.java#L2)):*
> ```Java
> /*
>  * Copyright (C) 2017 Google Inc.
> ```

also update the year here



_Code quote:_
```Java
2017
```
___
*[`src/main/java/com/google/android/mobly/snippet/bundled/utils/DataHolder.java` line 23 at r3](https://reviewable.io/reviews/google/mobly-bundled-snippets/167#-NWs4WnXDsllI4ZX3XCp:-NWsBfIbAHcoKOA_O9ps:b-n85y6d) ([raw file](https://github.com/google/mobly-bundled-snippets/blob/cc7a7d918f67ad4b2640b04aed84256c5fd316da/src/main/java/com/google/android/mobly/snippet/bundled/utils/DataHolder.java#L23)):*
<details><summary><i>Previously, ko1in1u (Kolin Lu) wrote…</i></summary><blockquote>

This is also a placeholder for future extensions between Snippet classes and Utils (like `JsonDeserializer.java`)
</blockquote></details>

But today it can only be used for BluetoothGattCharacteristic. If this behavior will change in the future maybe leave a TODO?

**Comment by ko1in1u** [comment]
> > _[Reviewable](https://reviewable.io/reviews/google/mobly-bundled-snippets/167)_ status: 4 of 10 files reviewed, 4 unresolved discussions (waiting on @ko1in1u and @xpconanfan)
> 
> _[`src/main/java/com/google/android/mobly/snippet/bundled/utils/DataHolder.java` line 2 at r3](https://reviewable.io/reviews/google/mobly-bundled-snippets/167#-NWsBw1l9XaMKwT4jSOB:-NWsBw1l9XaMKwT4jSOC:b-4asxxb) ([raw file](https://github.com/google/mobly-bundled-snippets/blob/cc7a7d918f67ad4b2640b04aed84256c5fd316da/src/main/java/com/google/android/mobly/snippet/bundled/utils/DataHolder.java#L2)):_
> 
> > ```java
> > /*
> >  * Copyright (C) 2017 Google Inc.
> > ```
> 
> also update the year here
> 
> _Code quote:_
> 
> ```java
> 2017
> ```

Updated.

> _[`src/main/java/com/google/android/mobly/snippet/bundled/utils/DataHolder.java` line 23 at r3](https://reviewable.io/reviews/google/mobly-bundled-snippets/167#-NWs4WnXDsllI4ZX3XCp:-NWsBfIbAHcoKOA_O9ps:b-n85y6d) ([raw file](https://github.com/google/mobly-bundled-snippets/blob/cc7a7d918f67ad4b2640b04aed84256c5fd316da/src/main/java/com/google/android/mobly/snippet/bundled/utils/DataHolder.java#L23)):_
> 
> Previously, ko1in1u (Kolin Lu) wrote…
> But today it can only be used for BluetoothGattCharacteristic. If this behavior will change in the future maybe leave a TODO?

Added, thanks.

**Review by xianyuanjia — APPROVED** [review]
:lgtm:

> *[Reviewable](https://reviewable.io/reviews/google/mobly-bundled-snippets/167#-:-NWsHFy48B20WIV2nCqe:bnfp4nl)* status: 4 of 10 files reviewed, 4 unresolved discussions (waiting on @ko1in1u and @xpconanfan)

**Pull request merged**